share/manual/sift.1
.TH sift "1" "Mar 2025" "sift 0.1" "sift manual"

.SH NAME
sift \- sift raw data into typed value(s)

.SH SYNOPSIS
\fBsift\fR [\fIOPTIONS\fR]

.SH DESCRIPTION
.PP
Sift raw data into an array of one or more typed values.

.TP
\fB\-t\fR \fITYPE\fR
value type (see \fBTYPES\fR for a list of supported types)

.TP
\fB\-f\fR \fIFORMAT\fR
value format string (see \fBTYPES\fR for default format strings)

.TP
\fB\-s\fR \fISIZE\fR
value size in bytes (default if unspecified: \fBsizeof(TYPE)\fR)

.TP
\fB\-c\fR \fICOUNT\fR
number of values in the array (default if unspecified: \fB1\fR)

.SH TYPES
.PP
Be aware that \fBsift\fR does not offer macro expansion at run time. Consult \fI<inttypes.h>\fR to determine how the PRI* macros listed below are expanded by the target's preprocessor at compile time.

.TP
\fBint8\fR
.RS
8-bit signed integer
.RE
.RS
"%" PRId8
.RE

.TP
\fBuint8\fR
.RS
8-bit unsigned integer
.RE
.RS
"%" PRIu8
.RE

.TP
\fBint16le\fR
.RS
16-bit little endian signed integer
.RE
.RS
"%" PRId16
.RE

.TP
\fBint16be\fR
.RS
16-bit big endian signed integer
.RE
.RS
"%" PRId16
.RE

.TP
\fBint16\fR
.RS
16-bit signed integer
.RE
.RS
"%" PRId16
.RE

.TP
\fBuint16le\fR
.RS
16-bit little endian unsigned integer
.RE
.RS
"%" PRIu16
.RE

.TP
\fBuint16be\fR
.RS
16-bit big endian unsigned integer
.RE
.RS
"%" PRIu16
.RE

.TP
\fBuint16\fR
.RS
16-bit unsigned integer
.RE
.RS
"%" PRIu16
.RE

.TP
\fBint32le\fR
.RS
32-bit little endian signed integer 
.RE
.RS
"%" PRId32
.RE

.TP
\fBint32be\fR
.RS
32-bit big endian signed integer 
.RE
.RS
"%" PRId32
.RE

.TP
\fBint32\fR
.RS
32-bit signed integer 
.RE
.RS
"%" PRId32
.RE

.TP
\fBuint32le\fR
.RS
32-bit little endian unsigned integer 
.RE
.RS
"%" PRIu32
.RE

.TP
\fBuint32be\fR
.RS
32-bit big endian unsigned integer 
.RE
.RS
"%" PRIu32
.RE

.TP
\fBuint32\fR
.RS
32-bit unsigned integer 
.RE
.RS
"%" PRIu32
.RE

.TP
\fBint64le\fR
.RS
64-bit little endian signed integer 
.RE
.RS
"%" PRId64
.RE

.TP
\fBint64be\fR
.RS
64-bit big endian signed integer 
.RE
.RS
"%" PRId64
.RE

.TP
\fBint64\fR
.RS
64-bit signed integer 
.RE
.RS
"%" PRId64
.RE

.TP
\fBuint64le\fR
.RS
64-bit little endian unsigned integer 
.RE
.RS
"%" PRIu64
.RE

.TP
\fBuint64be\fR
.RS
64-bit big endian unsigned integer 
.RE
.RS
"%" PRIu64
.RE

.TP
\fBuint64\fR
.RS
64-bit unsigned integer 
.RE
.RS
"%" PRIu64
.RE

.TP
\fBstring\fR
.RS
variable length NUL-terminated string
.RE
.RS
"%s"
.RE

.SH EXAMPLES

.SS Display a little endian 16-bit unsigned integer 
.BR
.EX
  $ printf "\\001\\000" | sift -t uint16le
  1
.EE

.SS Display a big endian 16-bit signed integer 
.BR
.EX
  $ printf "\\377\\377" | sift -t int16be
  -1
.EE

.SS Display a big endian 32-bit unsigned integer in hexadecimal
.BR
.EX
  $ printf "\\001\\000\\000\\000" | sift -f "%x" -t uint32be
  1000000
.EE

.SS Display an array of 4 x 8-bit unsigned integers
.BR
.EX
  $ printf "\\374\\375\\376\\377" | sift -c 4 -t uint8
  252
  253
  254
  255
.EE

.SS Display a native ELF64 object's \.shstrtab
.BR
.EX
Offset of the Section Header Table

  $ dump -o 0x28 -s 8 /bin/sh | sift -f "0x%x" -t uint64
  0x1e378


Size of a Section Header

  $ dump -o 0x3a -s 2 /bin/sh | sift -t uint16
  64


Index of the Section Name Table header

  $ dump -o 0x3e -s 2 /bin/sh | sift -t uint16
  27


Offset of \.shstrtab data

  $ dump -o $((0x1e378 + (64 * 27) + 0x18)) -s 8 /bin/sh | \\
    sift -f "0x%x" -t uint64
  0x1e274


Size of \.shstrtab data

  $ dump -o $((0x1e378 + (64 * 27) + 0x20)) -s 8 /bin/sh | \\
    sift -t uint64
  257


Contents of \.shstrtab

  $ dump -o 0x1e274 -s 257 /bin/sh | sift -c 27 -t string

  .shstrtab
  .interp
  .note.gnu.build-id
  .note.ABI-tag
  .gnu.hash
  .dynsym
  .dynstr
  .gnu.version
  .gnu.version_r
  .rela.dyn
  .rela.plt
  .init
  .plt.got
  .text
  .fini
  .rodata
  .eh_frame_hdr
  .eh_frame
  .init_array
  .fini_array
  .data.rel.ro
  .dynamic
  .data
  .bss
  .gnu_debuglink
.EE

.SH AUTHOR
Justin Swartz <justin.swartz@risingedge.co.za>

.SH SEE ALSO
dump(1)


src/sift.c
#include <errno.h>
#include <inttypes.h>
#include <limits.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

enum endianness {
	little,
	big
};

static enum endianness endianness;

enum key
{
	none,
	int8,
	uint8,
	int16le,
	int16be,
	int16,
	uint16le,
	uint16be,
	uint16,
	int32le,
	int32be,
	int32,
	uint32le,
	uint32be,
	uint32,
	int64le,
	int64be,
	int64,
	uint64le,
	uint64be,
	uint64,
	string,
};

struct type
{
	enum key key;
	size_t size;
	char *label;
	char *format;
};

static struct type types[] = {
	{ none,      0, NULL,        NULL       },
	{ int8,      1, "int8",      "%" PRId8  },
	{ uint8,     1, "uint8",     "%" PRIu8  },
	{ int16le,   2, "int16le",   "%" PRId16 },
	{ int16be,   2, "int16be",   "%" PRId16 },
	{ int16,     2, "int16",     "%" PRId16 },
	{ uint16le,  2, "uint16le",  "%" PRIu16 },
	{ uint16be,  2, "uint16be",  "%" PRIu16 },
	{ uint16,    2, "uint16",    "%" PRIu16 },
	{ int32le,   4, "int32le",   "%" PRId32 },
	{ int32be,   4, "int32be",   "%" PRId32 },
	{ int32,     4, "int32",     "%" PRId32 },
	{ uint32le,  4, "uint32le",  "%" PRIu32 },
	{ uint32be,  4, "uint32be",  "%" PRIu32 },
	{ uint32,    4, "uint32",    "%" PRIu32 },
	{ int64le,   8, "int64le",   "%" PRId64 },
	{ int64be,   8, "int64be",   "%" PRId64 },
	{ int64,     8, "int64",     "%" PRId64 },
	{ uint64le,  8, "uint64le",  "%" PRIu64 },
	{ uint64be,  8, "uint64be",  "%" PRIu64 },
	{ uint64,    8, "uint64",    "%" PRIu64 },
	{ string,    0, "string",    "%s"       }
};

static size_t typeCount = sizeof(types) / sizeof(*types);

static struct {
	size_t size;
	size_t count;
	struct type *type;
	char *format;
} config;

static void usage(char *program)
{
	fprintf(stderr, "usage: %s [OPTIONS]\n"
	                "\n"
	                "OPTIONS\n"
	                "\n"
	                "  -c COUNT\n"
	                "  -s SIZE\n"
	                "  -f FORMAT\n"
	                "  -t TYPE\n"
	                "\n"
	                "TYPES\n"
	                "\n"
	                "   int8\n"
	                "  uint8\n"
	                "   int16   int16le   int16be\n"
	                "  uint16  uint16le  uint16be\n"
	                "   int32   int32le   int32be\n"
	                "  uint32  uint32le  uint32be\n"
	                "   int64   int64le   int64be\n"
	                "  uint64  uint64le  uint64be\n"
	                "  string\n"
	                "\n", program);
}

static enum endianness testEndianness(void)
{
	static union {
		unsigned char data[sizeof(unsigned int)];
		unsigned int  result;
	} test;

	test.data[0] = UCHAR_MAX;
	if (test.result == test.data[0])
		return little;

	return big;
}

static uintmax_t parseNumber(char *string, int *status)
{
	char *terminator;
	uintmax_t number;
	
	errno = 0;
	number = strtoumax(string, &terminator, 0);
	
	if (errno) {
		perror("strtoumax");
		*status = -1;
		return 0;
	}
	
	if (*terminator != 0) {
		*status = -1;
		return 0;
	}

	return number;
}

static int parseArguments(int argc, char *argv[])
{
	int status = 0;

	config.count = 1;
	config.type  = types + none;

	if (argc < 2) {
		usage(argv[0]);
		return -1;
	}

	for (int index = 1; index < argc; index += 2) {
		char *option   = argv[index];
		char *argument = argv[index + 1];

		if (strcmp(option, "-c") == 0) {
			config.count = parseNumber(argument, &status);
			if (status == -1)
				return -1;
		}

		if (strcmp(option, "-s") == 0) {
			config.size = parseNumber(argument, &status);
			if (status == -1)
				return -1;
		}

		else if (strcmp(option, "-f") == 0) {
			config.format = argument;
		}

		else if (strcmp(option, "-t") == 0) { 
			for (size_t key = int8; key < typeCount; key++) {
				struct type *type = types + key;

				if (strcmp(argument, type->label) == 0) {
					config.type = type;

					if (config.size == 0)
						config.size = type->size;

					if (config.format == NULL)
						config.format = type->format;

					break;
				}
			}
		}
	}

	if (config.type == types + none) {
		usage(argv[0]);
		return -1;
	}

	return 0;
}

static int printString(void)
{
	char buffer[BUFSIZ];

	size_t remaining;
	size_t index = 0;

	while (!feof(stdin)) {
		char *cursor, *terminator;

		remaining = fread(buffer, 1, sizeof(buffer) - 1, stdin);
		if (remaining < 1) {
			if (ferror(stdin)) {
				perror("fread");
				return -1;
			}
		}

		cursor = buffer;
		terminator = buffer + remaining;
		*terminator = 0;

		while (cursor < terminator) {
			printf(config.format, cursor);

			if (cursor + strlen(cursor) == terminator)
				break;

			putchar('\n');

			if (++index == config.count)
				return 0;

			cursor = strchr(cursor, 0);
			if (cursor == NULL)
				return -1;

			cursor++;
		}
	}

	return -1;
}

static void reverseNumber(uintmax_t *number, size_t size)
{
	unsigned char *data = (unsigned char *)number;

	for (size_t index = 0; index < size / 2; index++) {
		unsigned char aux = data[index] ^ data[size - index - 1];
		data[index] ^= aux;
		data[size - index - 1] ^= aux;
	}
}

static int printNumber(void)
{
	union {
		int8_t    int8;
		uint8_t   uint8;
		int16_t   int16le;
		int16_t   int16be;
		int16_t   int16;
		uint16_t  uint16le;
		uint16_t  uint16be;
		uint16_t  uint16;
		int32_t   int32le;
		int32_t   int32be;
		int32_t   int32;
		uint32_t  uint32le;
		uint32_t  uint32be;
		uint32_t  uint32;
		int64_t   int64le;
		int64_t   int64be;
		int64_t   int64;
		uint64_t  uint64le;
		uint64_t  uint64be;
		uint64_t  uint64;
		uintmax_t uintmax;
	} number;

	number.uintmax = 0;

	for (size_t index = 0; index < config.count; index++) {
		struct type *type = config.type;

		if (fread(&number, config.size, 1, stdin) < 1) {
			if (ferror(stdin))
				perror("fread");
			return -1;
		}

		switch (type->key) {
		case int16le:
		case uint16le:
			if (endianness == big)
				reverseNumber(&number.uintmax, 2);
			break;

		case int16be:
		case uint16be:
			if (endianness == little)
				reverseNumber(&number.uintmax, 2);
			break;

		case int32le:
		case uint32le:
			if (endianness == big)
				reverseNumber(&number.uintmax, 4);
			break;

		case int32be:
		case uint32be:
			if (endianness == little)
				reverseNumber(&number.uintmax, 4);
			break;

		case int64le:
		case uint64le:
			if (endianness == big)
				reverseNumber(&number.uintmax, 8);
			break;

		case int64be:
		case uint64be:
			if (endianness == little)
				reverseNumber(&number.uintmax, 8);
			break;

		default:
			break;
		}

		switch (type->key) {
		case int8:
			printf(config.format, number.int8);
			break;

		case uint8:
			printf(config.format, number.uint8);
			break;

		case int16le:
			printf(config.format, number.int16le);
			break;

		case int16be:
			printf(config.format, number.int16be);
			break;

		case int16:
			printf(config.format, number.int16);
			break;

		case uint16le:
			printf(config.format, number.uint16le);
			break;

		case uint16be:
			printf(config.format, number.uint16be);
			break;

		case uint16:
			printf(config.format, number.uint16);
			break;

		case int32le:
			printf(config.format, number.int32le);
			break;

		case int32be:
			printf(config.format, number.int32be);
			break;

		case int32:
			printf(config.format, number.int32);
			break;

		case uint32le:
			printf(config.format, number.uint32le);
			break;

		case uint32be:
			printf(config.format, number.uint32be);
			break;

		case uint32:
			printf(config.format, number.uint32);
			break;

		case int64le:
			printf(config.format, number.int64le);
			break;

		case int64be:
			printf(config.format, number.int64be);
			break;

		case int64:
			printf(config.format, number.int64);
			break;

		case uint64le:
			printf(config.format, number.uint64le);
			break;

		case uint64be:
			printf(config.format, number.uint64be);
			break;

		case uint64:
			printf(config.format, number.uint64);
			break;

		default:
			break;
		}

		putchar('\n');
	}

	return 0;
}

static int format(void)
{
	struct type *type = config.type;

	switch (type->key) {
	case string:
		if (printString() == -1)
			return -1;
		break;

	default:
		if (printNumber() == -1)
			return -1;
		break;
	}

	return 0;
}

int main(int argc, char *argv[])
{
	endianness = testEndianness();

	if (parseArguments(argc, argv) == -1)
		return EXIT_FAILURE;

	if (format() == -1)
		return EXIT_FAILURE;

	return EXIT_SUCCESS;
}
